url8f84d8d956_klapaukh_JTuttePoly_tgz-pJ8-BitTutteJ8/src/graph/AdjacencyMatrix.java

```diff
@@ -0,0 +1,931 @@
+	 */
+	public void clear(int v) {
+		// Remove v from the list of vertices
+		if (startVertex == v) {
+			startVertex = vertices[v];
+		} else {
+			int i = startVertex;
+			while (vertices[i] != v) {
+				i = vertices[i];
+				if (i == -1) {
+					return;
+				}
+			}
+			vertices[i] = vertices[v];
+		}
+		vertices[v] = -1;
+		numVertices--;
+
+		// first delete the edges that don't start at you
+		// Now delete the columns
+		for (int i = 0; i < domainSize; i++) {
+			int num = numEdges(i, v);
+			if (num != 0) {
+				numEdges -= num;
+				numMultiEdges -= (num - 1);
+				setValue(i, v, 0);
+			}
+		}
+
+		// Now, clear all edges involving v
+		// first the row This is done long ways as this might be faster as it
+		// can write an int at a time
+		long startBit = domainSize * CELL_SIZE * v;
+		long endBit = domainSize * CELL_SIZE * (v + 1);
+
+		int startInt = (int) (startBit / 32);
+		int endInt = (int) (endBit / 32);
+		int endIndex = (int) (endBit % 32);
+		int startIndex = (int) (startBit % 32);
+
+		if (startInt == endInt) {
+			// If the whole row is contain within the current int
+			int m = 0xFFFFFFFF << (domainSize * CELL_SIZE);
+			m |= (1 << startIndex) - 1;
+			edges[startInt] &= m;
+		} else {
+			// clear the starting int
+			if (startIndex != 0) {
+				// Need to only clear the END of the starting int
+				int m = (1 << startIndex) - 1;
+				edges[startInt] &= m;
+				startInt++;
+			}
+			while (startInt < endInt) {
+				edges[startInt++] = 0;
+			}
+			// Clear as much of the last index as needed
+			if (endIndex != 0) {
+				int m = 0xFFFFFFFF << endIndex;
+				edges[endInt] &= m;
+			}
+
+		}
+
+	}
+
+	private void setValue(int from, int to, int val) {
+		setValue(from,to, val, domainSize,edges);
+	}
+
+	/**
+	 * Add c many of an edge from from to to
+	 * 
+	 * @param from
+	 * @param to
+	 * @param c
+	 * @return True if the edge already exists
+	 */
+	public boolean addEdge(int from, int to, int c) {
+		numEdges += c;
+
+		int num = this.numEdges(from, to);
+		if (num == 0) {
+			numMultiEdges += c - 1;
+		} else {
+			numMultiEdges += c;
+		}
+
+		c += num;
+
+		setValue(from, to, c);
+		if (from != to) {
+			setValue(to, from, c);
+		}
+
+		return num != 0;
+	}
+
+	/**
+	 * Add an undirected edge from from to to
+	 * 
+	 * @param from
+	 *            Source
+	 * @param to
+	 *            Sink
+	 * @return True if the edge already exists
+	 */
+	public boolean addEdge(int from, int to) {
+		return addEdge(from, to, 1);
+	}
+
+	public boolean removeEdge(int from, int to, int c) {
+
+		int i = numEdges(from, to);
+		if (i == 0) {
+			return false;
+		}
+		if (i > c) {
+			// this is a multi-edge, so decrement count.
+			numMultiEdges -= c;
+			numEdges -= c;
+			setValue(from, to, i - c);
+			if (from != to) {
+				setValue(to, from, i - c);
+			}
+		} else {
+			// set to zero
+			numEdges -= i;
+			numMultiEdges -= (i - 1);
+			setValue(from, to, 0);
+			if (from != to) {
+				setValue(to, from, 0);
+			}
+		}
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 322:330, 337:344.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.